restart_main.cpp uses SIZE instead of MIN_SIZE

In some places in `restart_main.cpp` the constant `SIZE` is referenced but not defined which leads to a compilation error like this one:
```
restart_main.cpp: In function ‘int main(int, char**)’:
restart_main.cpp:105:16: error: ‘SIZE’ was not declared in this scope
  if(data.len > SIZE) data.len = SIZE;
                ^~~~
restart_main.cpp:105:16: note: suggested alternative: ‘FILE’
  if(data.len > SIZE) data.len = SIZE;
                ^~~~
                FILE
restart_main.cpp:107:16: error: ‘SIZE’ was not declared in this scope
  if(data.len < SIZE){
                ^~~~
restart_main.cpp:107:16: note: suggested alternative: ‘FILE’
  if(data.len < SIZE){
                ^~~~
                FILE
```
My guess is that this should be changed to `MIN_SIZE` as this is defined in the file `utils.h` and used in the implementation of iid and non_iid.

--- a/README.md
+++ b/README.md
@@ -14,7 +14,9 @@ Issues on this repository are strictly for problems or questions concerning the 
 
 ## Requirements
 
-This code package requires a C++11 compiler. GCC is preferred (and the only platform tested). There is one method that involves a GCC-only method (`chi_square_tests.h -> binary_goodness_of_fit() -> __builtin_popcount()`). To run this you will need GCC.
+This code package requires a C++11 compiler. The code uses OpenMP directives, so compiler support for OpenMP is expected. GCC is preferred (and the only platform tested). There is one method that involves a GCC built-in function (`chi_square_tests.h -> binary_goodness_of_fit() -> __builtin_popcount()`). To run this you will need some compiler that supplies this GCC built-in function (GCC and clang both do so).
+
+The resulting binary is linked with bzlib, so this library (and its associated include files) must be installed and accessible to the compiler.
 
 ## Overview
 

--- a/cpp/Makefile
+++ b/cpp/Makefile
@@ -1,17 +1,16 @@
 CXX = g++
 CXXFLAGS = -std=c++11 -fopenmp
-CDDFLAGS = -std=c++11 -ggdb -fopenmp
-TESTFLAGS = -ftree-vectorize -msse2 -Ofast
-
 LIB = -lbz2 -lpthread
-INC=-I/usr/include -I/usr/include/c++/4.9 -I/usr/include/c++/4.9/backward \
- -I/usr/include/x86_64-linux-gnu -I/usr/include/x86_64-linux-gnu/c++/4.9 \
- -I/usr/lib/gcc/x86_64-linux-gnu/4.9/include -I/usr/lib/gcc/x86_64-linux-gnu/4.9/include-fixed \
- -I/usr/local/include
+INC=
 
 ######
 # Main operations
 ######
+
+all:    iid non_iid restart conditioning
+
+clean:
+	rm -f ea_iid ea_non_iid ea_restart ea_conditioning
 
 iid: iid_main.o
 iid_main.o: iid_main.cpp
@@ -21,41 +20,16 @@ non_iid: non_iid_main.o
 non_iid_main.o: non_iid_main.cpp
 	$(CXX) $(CXXFLAGS) $(LIB) $(INC) non_iid_main.cpp -o ea_non_iid
 
+restart: restart_main.o	
+restart_main.o: restart_main.cpp
+	$(CXX) $(CXXFLAGS) $(LIB) $(INC) restart_main.cpp -o ea_restart
+
+conditioning: conditioning_main.o
+conditioning_main.o: conditioning_main.cpp
+	$(CXX) $(CXXFLAGS) $(LIB) $(INC) conditioning_main.cpp -o ea_conditioning
+
 run_iid:
 	./ea_iid ../bin/truerand_8bit.bin 8 -i -t -v
 
 run_non_iid:
 	./ea_non_iid ../bin/truerand_8bit.bin 8 -i -t -v
-
-
-######
-# Test operations
-######
-
-chi_test: chi_test_main.o
-chi_test_main.o: test/chi_test_main.cpp
-	$(CXX) $(CXXFLAGS) $(TESTFLAGS) $(INC) test/chi_test_main.cpp $(LIB) -o chi_test
-
-run_chi_test:
-	./chi_test
-
-lrs_test: lrs_test_main.o
-lrs_test_main.o: test/lrs_test_main.cpp
-	$(CXX) $(CXXFLAGS) $(TESTFLAGS) $(INC) test/lrs_test_main.cpp $(LIB) -o lrs_test
-
-run_lrs_test:
-	./lrs_test
-
-most_common_test: most_common_test_main.o
-most_common_test_main.o: test/most_common_test_main.cpp
-	$(CXX) $(CXXFLAGS) $(TESTFLAGS) $(INC) test/most_common_test_main.cpp $(LIB) -o most_common_test
-
-run_most_common_test:
-	./most_common_test
-
-permutation_test: permutation_test_main.o
-permutation_test_main.o: test/permutation_test_main.cpp
-	$(CXX) $(CXXFLAGS) $(TESTFLAGS) $(INC) test/permutation_test_main.cpp $(LIB) -o permutation_test
-
-run_permutation_test:
-	./permutation_test ../bin/truerand_8bit.bin 8 1000000 1 -v

--- a/cpp/conditioning_main.cpp
+++ b/cpp/conditioning_main.cpp
@@ -1,19 +1,11 @@
-#include "shared/utils.h"
-#include "shared/most_common.h"
-#include "shared/lrs_test.h"
-#include "non_iid/collision_test.h"
-#include "non_iid/lz78y_test.h"
-#include "non_iid/multi_mmc_test.h"
-#include "non_iid/lag_test.h"
-#include "non_iid/multi_mcw_test.h"
-#include "non_iid/tuple.h"
-#include "non_iid/compression_test.h"
-#include "non_iid/markov_test.h"
-
+#include <stdio.h>
+#include <cstdlib>
 #include <limits>
+#include <cmath>
+#include <algorithm>
 
 void print_usage(){
-	printf("Usage is: ./conditioning_main <n_in> <n_out> <nw> <h_in> <-v|-n> <h'>\n\n");
+	printf("Usage is: ea_conditioning <n_in> <n_out> <nw> <h_in> <-v|-n> <h'>\n\n");
 	printf("\t <n_in>: input number of bits to conditioning function.\n");
 	printf("\t <n_out>: output number of bits from conditioning function.\n");
 	printf("\t <nw>: narrowest internal width of conditioning function.\n");
@@ -31,11 +23,8 @@ void print_usage(){
 }
 
 int main(int argc, char* argv[]){
-	bool vetted, verbose = false;
-	const char verbose_flag = 'v';
-	char *file_path;
-	double h_p, h_in, h_out, n_in, n_out, nw, n, p_high, p_low, psi, omega, output_entropy;
-	data_t data;
+	bool vetted;
+	double h_p, h_in, h_out, n_in, n_out, nw, n, p_high, p_low, psi, omega, output_entropy, power_term;
 
 	// Parse args
 	if(argc != 6 && argc != 7){
@@ -103,19 +92,24 @@ int main(int argc, char* argv[]){
 		}
 	}
 
+	if(nw > n_in) nw = n_in;
+
 	printf("n_in: %f\n", n_in);
 	printf("n_out: %f\n", n_out);
 	printf("nw: %f\n", nw);
 	printf("h_in: %f\n", h_in);
 	if(!vetted) printf("h': %f\n", h_p);
-	
+
 	// compute Output Entropy (Section 3.1.5.1.2)
 	p_high = pow(2.0, -h_in);
 	p_low = (1.0-p_high)/(pow(2.0, n_in)-1);
-	n = min(n_out, nw);
-	psi = pow(2.0, n_in-n)*p_low + p_high;
-	omega = (pow(2.0, n_in-n) + sqrt(2*n*(pow(2.0, n_in-n)*log(2))))*p_low;
-	output_entropy = -log2(max(psi, omega));
+	n = std::min(n_out, nw);
+	power_term = pow(2.0, n_in - n);
+	psi = power_term*p_low + p_high;
+	omega = (power_term + sqrt(2*n*power_term*log(2)))*p_low;
+	output_entropy = -log2(std::max(psi, omega));
+	if(output_entropy > n_out) output_entropy = n_out;
+	if(output_entropy < 0) output_entropy = 0;
 
 	if(vetted){
 		printf("\n(Vetted) ");
@@ -123,7 +117,7 @@ int main(int argc, char* argv[]){
 	}
 	else{
 		printf("\n(Non-vetted) ");
-		h_out = min(output_entropy, min(0.999*n_out, h_p*n_out));
+		h_out = std::min(output_entropy, std::min(0.999*n_out, h_p*n_out));
 	}
 
 	printf("h_out: %f\n", h_out);

--- a/cpp/iid_main.cpp
+++ b/cpp/iid_main.cpp
@@ -5,7 +5,7 @@
 #include "iid/chi_square_tests.h"
 
 void print_usage(){
-	printf("Usage is: ./iid_main.out <file_name> <bits_per_word> <-i|-c> <-a|-t> [-v]\n\n");
+	printf("Usage is: ea_iid <file_name> <bits_per_word> <-i|-c> <-a|-t> [-v]\n\n");
 	printf("\t <file_name>: Must be relative path to a binary file with at least 1 million entries (words).\n");
 	printf("\t <bits_per_word>: Must be between 1-8, inclusive.\n");
 	printf("\t <-i|-c>: '-i' for initial entropy estimate, '-c' for conditioned sequential dataset entropy estimate.\n");

--- a/cpp/non_iid_main.cpp
+++ b/cpp/non_iid_main.cpp
@@ -12,7 +12,7 @@
 
 
 void print_usage(){
-	printf("Usage is: ./non_iid_main <file_name> <bits_per_word> <-i|-c> <-a|-t> [-v]\n\n");
+	printf("Usage is: ea_non_iid <file_name> <bits_per_word> <-i|-c> <-a|-t> [-v]\n\n");
 	printf("\t <file_name>: Must be relative path to a binary file with at least 1 million entries (words).\n");
 	printf("\t <bits_per_word>: Must be between 1-8, inclusive.\n");
 	printf("\t <-i|-c>: '-i' for initial entropy estimate, '-c' for conditioned sequential dataset entropy estimate.\n");

--- a/cpp/restart_main.cpp
+++ b/cpp/restart_main.cpp
@@ -12,7 +12,7 @@
 
 
 void print_usage(){
-	printf("Usage is: ./restart_main <file_name> <bits_per_word> <H_I> <-i|-n> [-v]\n\n");
+	printf("Usage is: ea_restart <file_name> <bits_per_word> <H_I> <-i|-n> [-v]\n\n");
 	printf("\t <file_name>: Must be relative path to a binary file with at least 1 million entries (words).\n");
 	printf("\t <bits_per_word>: Must be between 1-8, inclusive.\n");
 	printf("\t <H_I>: Initial entropy estimate.\n");
@@ -102,12 +102,11 @@ int main(int argc, char* argv[]){
                 exit(-1);
         }
 
-	if(data.len > SIZE) data.len = SIZE;
-	if(verbose) printf("Number of Symbols: %ld\n", data.len);
-	if(data.len < SIZE){ 
-		printf("\n*** Error: data contains less than %d samples ***\n\n", SIZE);
+	if(data.len != MIN_SIZE){ 
+		printf("\n*** Error: data does not contain %d samples ***\n\n", MIN_SIZE);
 		exit(-1);
 	}
+	if(verbose) printf("Number of Symbols: %ld\n", data.len);
 	if(verbose){
 		if(data.alph_size < (1 << data.word_size)) printf("\nSymbols have been mapped down to an alphabet size of %d unique symbols\n\n", data.alph_size);
 		else printf("\nSymbol alphabet consists of %d unique symbols\n\n", data.alph_size);

